### ISSUE TITLE
generic "linuxkit push scp" front end to scp to support iPXE based booting

### ISSUE THREAD

**Issue by vielmetti** [linuxkit/linuxkit#2049]
**Description**

Support for Linuxkit on Packet is missing a "push" action. In truth, there really isn't one that's specific to Packet; rather, it would be useful and meaningful to have a set of generic "push" operations that would upload the necessary artifacts to an arbitrary server.

Specifically, the Packet support does "linuxkit run" using iPXE, and all the "push" command needs to do is to upload files to a named server with a well-understood file transfer protocol. 

This would be generally useful for other providers, or even as a backup mechanism. Syntax noted below is a suggestion not a requirement.

**Steps to reproduce the issue:**

`linuxkit push` and notice that support is tied to specific platforms, hiding the mechanics of the operation, and doesn't support the general case of scripted file transfer of known artifacts to a named target.

**Describe the results you expected:**

`linuxkit push scp [options] [prefix]` should do the right thing, e.g. with a command line that accepts the scp command options

```
linuxkit push scp -i ~/.ssh/example.pub user@ipxe-host:/path/to/ipxe/ linuxkit
```

would do the same as

```
scp -i ~/.ssh/example.pub linuxkit-kernel user@ipxe-host:/path/to/ipxe/linuxkit-kernel
scp -i ~/.ssh/example.pub linuxkit-initrd.img user@ipxe-host:/path/to/ipxe/linuxkit-initrd.img
```

**Additional information you deem important (e.g. issue happens only occasionally):**

Naturally this same approach could be extended to other file transfer protocols.

This is mostly directly related to the issue previous ( #2046 ) since it took me a moment to decode what was needed.

**Comment by justincormack** [comment]
Seems reasonable. We could also have a `linuxkit push s3`, which should be pretty easy to add as we have all the code already in the `push aws`.

**Comment by vielmetti** [comment]
Yup, a `push s3` would be super useful - as it turns out the server I ended up using for my test pxe boot had `minio` installed, and I had an `mc` client, so what I actually did was the equivalent of

```
mc cp linuxkit-kernel ipxe-host/ipxe
mc cp linuxkit-initrd.img ipxe-host/ipxe
```

If you're doing an aws s3 client take care enough to make it also work on other s3-compatible systems.

**Comment by justincormack** [comment]
I havent used `minio` though have heard good things, so not sure how compatible it is with the AWS libraries (is the auth identical?). @alexellis you have used it fairly extensively havent you?

**Comment by rn** [comment]
another option would be use ngrok as shown here: https://www.youtube.com/watch?v=j50ovfRWpZM

about 8min in

**Comment by rn** [comment]
I added `linuxkit packet push` which does not yet have a `scp` style backend but creates the kernel/initrd/iPXE script in a directory which can then be copied to a server. I'm not sure we want `scp` support as this is rather tedious wrt to agent integration etc

**Comment by alexellis** [comment]
@justincormack I missed this - the Minio API is available in Go and is compatible with AWS. That would be a great option and I'd interested in helping out.

**Comment by rn** [comment]
@alexellis sure. I normally use ngrok and boot from my local machine but AWS/S3 would also be a good option. I'd be happy to look at patches. The `linuxkit packet push` code is already structure so that it can take a URL to push to so something like `s3://<blah>` would be easy to add.

My two (minor) concerns are: 
- Handling of credentials. Ideally It would follow the same pattern as the `linuxkit run/push aws` stuff.
- You need the public URL before pushing as it is embedded into the iXPE script. I don't know S3/minio well enough to know if that is possible.
